lib/active_publisher/async/in_memory_adapter/consumer_thread.rb

```diff
@@ -39,43 +39,36 @@ def start_thread
           return if alive?
           @thread = ::Thread.new do
             loop do
-              # Write "current_messages" so we can requeue should something happen to the consumer.
-              @current_messages.concat(queue.pop_up_to(20))
+              # Sample the queue size so we don't shutdown when messages are in flight.
+              @sampled_queue_size = queue.size
+              current_messages = queue.pop_up_to(20)
 
               begin
                 # Only open a single connection for each group of messages to an exchange
-                messages_to_retry = []
-                @current_messages.group_by(&:exchange_name).each do |exchange_name, messages|
+                current_messages.group_by(&:exchange_name).each do |exchange_name, messages|
                   begin
+                    current_messages -= messages
                     ::ActivePublisher.publish_all(exchange_name, messages)
                   ensure
-                    messages_to_retry.concat(messages)
+                    current_messages.concat(messages)
                   end
                 end
-
-                # Reset
-                @current_messages = []
               rescue *NETWORK_ERRORS
                 # Sleep because connection is down
                 await_network_reconnect
-                @current_messages.concat(messages_to_retry)
 
                 # Requeue and try again.
-                queue.concat(@current_messages) unless @current_messages.empty?
+                queue.concat(current_messages)
               rescue => unknown_error
-                @current_messages.concat(messages_to_retry)
-                @current_messages.each do |message|
+                current_messages.each do |message|
                   # Degrade to single message publish ... or at least attempt to
                   begin
                     ::ActivePublisher.publish(message.route, message.payload, message.exchange_name, message.options)
-                  rescue => error
+                  rescue
                     ::ActivePublisher.configuration.error_handler.call(unknown_error, {:route => message.route, :payload => message.payload, :exchange_name => message.exchange_name, :options => message.options})
                   end
                 end
 
-                # Do not requeue the message because something else horrible happened.
-                @current_messages = []
-
                 # TODO: Find a way to bubble this out of the thread for logging purposes.
                 # Reraise the error out of the publisher loop. The Supervisor will restart the consumer.
                 raise unknown_error
```

Comment:
I don't think I changed behavior, though because we were already reseting the `@current_messages` just above this, so when we raise an error, there's nothing to re-queue.